File: solutions/union-find/261-graph-valid-tree.py
I'll help you implement a solution for the Graph Valid Tree Union problem using Python. This solution will use a Union-Find data structure to determine if a given graph is a valid tree.

```python
from typing import List

class UnionFind:
    """
    Union-Find data structure implementation for detecting cycles and connectivity in graphs
    """
    def __init__(self, n: int):
        """Initialize Union-Find data structure with n nodes"""
        self.parent = list(range(n))  # Each node is initially its own parent
        self.rank = [0] * n  # Rank array for union by rank optimization
        
    def find(self, x: int) -> int:
        """Find the parent of node x with path compression"""
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]
    
    def union(self, x: int, y: int) -> bool:
        """
        Union two nodes x and y
        Returns False if nodes are already connected (indicating a cycle), True otherwise
        """
        px, py = self.find(x), self.find(y)
        
        if px == py:  # If nodes have same parent, there's a cycle
            return False
        
        # Union by rank
        if self.rank[px] < self.rank[py]:
            self.parent[px] = py
        elif self.rank[px] > self.rank[py]:
            self.parent[py] = px
        else:
            self.parent[py] = px
            self.rank[px] += 1
            
        return True

def validTree(n: int, edges: List[List[int]]) -> bool:
    """
    Determine if an undirected graph is a valid tree
    
    Args:
        n: Number of nodes (0 to n-1)
        edges: List of edges where each edge is [node1, node2]
        
    Returns:
        bool: True if the graph is a valid tree, False otherwise
    """
    # A tree with n nodes must have exactly n-1 edges
    if len(edges) != n - 1:
        return False
    
    # Initialize Union-Find data structure
    uf = UnionFind(n)
    
    # Process each edge
    for x, y in edges:
        # If union returns False, we've found a cycle
        if not uf.union(x, y):
            return False
    
    return True

def test_valid_tree():
    """Test cases for validTree function"""
    # Test case 1: Valid tree
    assert validTree(5, [[0,1], [0,2], [0,3], [1,4]]) == True
    print("Test case 1 passed")
    
    # Test case 2: Graph with cycle
    assert validTree(5, [[0,1], [1,2], [2,3], [1,3], [1,4]]) == False
    print("Test case 2 passed")
    
    # Test case 3: Disconnected graph
    assert validTree(4, [[0,1], [2,3]]) == False
    print("Test case 3 passed")
    
    # Test case 4: Single node
    assert validTree(1, []) == True
    print("Test case 4 passed")
    
    # Test case 5: Two nodes connected
    assert validTree(2, [[0,1]]) == True
    print("Test case 5 passed")

if __name__ == "__main__":
    # Run test cases
    test_valid_tree()
    print("All test cases passed!")
```

This implementation includes:

1. A `UnionFind` class that implements the Union-Find data structure with:
   - Path compression optimization
   - Union by rank optimization
   - Methods for finding parents and unioning nodes

2. A `validTree` function that determines if a graph is a valid tree by:
   - Checking if the number of edges is correct (n-1 for n nodes)
   - Using Union-Find to detect cycles
   - Ensuring all nodes are connected

3. Test cases covering various scenarios:
   - Valid tree structures
   - Graphs with cycles
   - Disconnected graphs
   - Edge cases (single node, two nodes)

The solution efficiently handles the requirements by:
- Using optimal Union-Find implementation with path compression and union by rank
- Including proper type hints and documentation
- Following Python conventions and best practices
- Handling edge cases appropriately
- Providing comprehensive test cases

The code is complete and ready to run, with clear structure and comments explaining the implementation details.

File: solutions/union-find/261-graph-valid-tree.js
I'll help you create a JavaScript implementation for the Graph Valid Tree Union problem. This solution will determine if a given undirected graph forms a valid tree using the Union-Find data structure.

```javascript
/**
 * Graph Valid Tree Union Implementation
 * 
 * Problem: Given n nodes labeled from 0 to n-1 and a list of undirected edges,
 * determine if these edges form a valid tree.
 * 
 * A valid tree has:
 * 1. No cycles
 * 2. All nodes are connected
 * 3. Number of edges = number of nodes - 1
 */

class UnionFind {
    constructor(n) {
        // Initialize parent array where each node is its own parent
        this.parent = Array(n).fill(0).map((_, i) => i);
        // Initialize rank array for union by rank optimization
        this.rank = Array(n).fill(0);
    }

    /**
     * Find the root parent of a node with path compression
     * @param {number} x - The node to find the parent for
     * @returns {number} The root parent of the node
     */
    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]); // Path compression
        }
        return this.parent[x];
    }

    /**
     * Union two nodes by rank
     * @param {number} x - First node
     * @param {number} y - Second node
     * @returns {boolean} True if union was successful (no cycle), false if cycle detected
     */
    union(x, y) {
        let rootX = this.find(x);
        let rootY = this.find(y);

        // If roots are same, cycle exists
        if (rootX === rootY) {
            return false;
        }

        // Union by rank
        if (this.rank[rootX] < this.rank[rootY]) {
            [rootX, rootY] = [rootY, rootX]; // Swap to ensure rootX has higher rank
        }
        
        this.parent[rootY] = rootX;
        if (this.rank[rootX] === this.rank[rootY]) {
            this.rank[rootX]++;
        }

        return true;
    }
}

/**
 * Determines if a graph is a valid tree
 * @param {number} n - Number of nodes
 * @param {number[][]} edges - Array of edges where each edge is [node1, node2]
 * @returns {boolean} True if the graph forms a valid tree, false otherwise
 */
function validTree(n, edges) {
    // Check if number of edges is correct for a tree
    if (edges.length !== n - 1) {
        return false;
    }

    // Initialize Union-Find data structure
    const uf = new UnionFind(n);

    // Process each edge
    for (const [node1, node2] of edges) {
        // If union returns false, we found a cycle
        if (!uf.union(node1, node2)) {
            return false;
        }
    }

    return true;
}

// Test cases
const testCases = [
    {
        n: 5,
        edges: [[0,1], [0,2], [0,3], [1,4]],
        expected: true
    },
    {
        n: 5,
        edges: [[0,1], [1,2], [2,3], [1,3], [1,4]],
        expected: false
    },
    {
        n: 4,
        edges: [[0,1], [2,3]],
        expected: false
    }
];

// Run test cases
testCases.forEach((test, index) => {
    const result = validTree(test.n, test.edges);
    console.log(`Test case ${index + 1}:`);
    console.log(`Input: n = ${test.n}, edges = ${JSON.stringify(test.edges)}`);
    console.log(`Expected: ${test.expected}`);
    console.log(`Result: ${result}`);
    console.log(`Status: ${result === test.expected ? 'PASSED' : 'FAILED'}`);
    console.log('---');
});

// Export for external use
module.exports = {
    validTree,
    UnionFind
};
```

This implementation includes:

1. A `UnionFind` class with:
   - Path compression optimization
   - Union by rank optimization
   - Methods for finding parents and unioning nodes

2. A `validTree` function that:
   - Checks if the number of edges is valid (n-1 for a tree)
   - Uses Union-Find to detect cycles
   - Returns whether the graph forms a valid tree

3. Test cases to verify the implementation

4. Proper error handling and edge cases

5. Clear comments explaining the code

6. Module exports for external use

The solution efficiently determines if a graph is a valid tree by checking:
- The correct number of edges (n-1)
- No cycles exist (using Union-Find)
- All nodes are connected (implicit in the edge count check)

The time complexity is O(α(n)) per union operation, where α is the inverse Ackermann function (effectively constant time). The space complexity is O(n) for the Union-Find data structure.